### ISSUE TITLE
Loading a layer that uses weights from another layer

### ISSUE THREAD

**Issue by lvapeab** [keras-team/keras#10485]
Hi,

I want to implement a layer that uses the weights from another layer. 
I was trying to implement the tied embeddings from [Press and Wolf](https://arxiv.org/pdf/1608.05859.pdf), but this issue also applies to other models (e.g. tied autoencoders).

In my case, I want a seq2seq model using the same word embeddings in the input and output. The `Embedding` input layer has the dimensions`V`(input) and `H` (output). I want my output layer to be the same, but transposed (from `H` to `V`). Something like this (red connections denote shared weights): 

![embeddings](https://user-images.githubusercontent.com/4717075/41644057-57ee43aa-746d-11e8-9985-467fbaae0f47.png)

I implemented it via a shared layers. My input is a shared Embedding layer. And I defined a `TiedEmbeddingsTransposed` layer, which transposes the embedding matrix from a given layer (and applies an activation function):

```
class TiedEmbeddingsTransposed(Layer):
    """Layer for tying embeddings in an output layer.
    A regular embedding layer has the shape: V x H (V: size of the vocabulary. H: size of the projected space).
    In this layer, we'll go: H x V.
    With the same weights than the regular embedding.
    In addition, it may have an activation.
    # References
        - [ Using the Output Embedding to Improve Language Models](https://arxiv.org/abs/1608.05859)
    """

    def __init__(self, tied_to=None,
                 activation=None,
                 **kwargs):
        super(TiedEmbeddingsTransposed, self).__init__(**kwargs)
        self.tied_to = tied_to
        self.activation = activations.get(activation)

    def build(self, input_shape):
        self.transposed_weights = K.transpose(self.tied_to.weights[0])
        self.built = True

    def compute_mask(self, inputs, mask=None):
        return mask

    def compute_output_shape(self, input_shape):
        return input_shape[0], K.int_shape(self.tied_to.weights[0])[0]

    def call(self, inputs, mask=None):
        output = K.dot(inputs, self.transposed_weights)
        if self.activation is not None:
            output = self.activation(output)
        return output


    def get_config(self):
        config = {'activation': activations.serialize(self.activation)
                  }
        base_config = super(TiedEmbeddingsTransposed, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))
```

The usage of this layer is something like this:

```
# Declare the shared embedding layer
shared_embedding_layer = Embedding(V, H)
# Obtain word embeddings
word_embedding = shared_embedding_layer(input)
# Do stuff with your model
# Compute output (e.g. a vocabulary-size probability vector) with the shared layer:
output = TimeDistributed(TiedEmbeddingsTransposed(tied_to=shared_embedding_layer, activation='softmax')(intermediate_rep)
```
I have tested this and it trains properly. But, as I try to load a trained model, it gets an error, related to the way Keras loads the models: it doesn't load the weights from the `tied_to` layer. I tried different ways for accessing the `tied_to` layer weights and saving it, but no one seems to work, I've found several questions regarding this, related to autoencoders ([1](https://stackoverflow.com/questions/42015116/decoders-weights-of-autoencoder-with-tied-weights-in-keras), [2](https://stackoverflow.com/questions/47106830/keras-get-weights-returning-empty), [3](https://gist.github.com/dswah/c6b3e4d47d933b057aab32c9c29c4221#gistcomment-2298473)), but it seems no one has managed to solve this issue. Maybe this shared-layer approach is not the way to go. I'd be glad to hear some ideas/criticisms from the Keras' community for addressing this issue :)

Thanks and cheers!

**Comment by lvapeab** [comment]
Hi,
this is the same error I faced. Unfortunately, I haven't found a solution to this yet :(

**Comment by tea1528** [comment]
Hi, I've found a similar issue that may help you.
#5060

The problem I had with your custom layer was that the `tied_to` layer wasn't being passed on to the saved model. I think that is the reason why `AttributeError` was thrown after `load_model()`. 

I've reimplemented the code for my own purpose but it also has some problem. For instance, when I try to save the model, pickle error occurs. I think it must be related to the issue with `get_config()` passing `other_layer` by deepcopy. 

As a workaround, I didn't save the model but just ran the inference right after training. For some strange reason, the performance didn't really improve for simple lanuage modeling code. Nonetheless, I hope you find this helpful. Thanks.

```
from keras import backend as K
from keras.layers import Layer, Dense
import keras.activations as activations
from keras.engine.base_layer import InputSpec

class DenseTranspose(Dense):
    def __init__(self, other_layer=None, activation=None, **kwargs):
        super().__init__(other_layer.input_dim, **kwargs)
        self.other_layer = other_layer
        self.activation = activations.get(activation)


    def build(self, input_shape):
        assert len(input_shape) >= 2
        input_dim = input_shape[-1]
        self.input_dim = input_dim
        self.input_spec = [InputSpec(dtype=K.floatx(), ndim='2+')]

        self.kernel = K.transpose(self.other_layer.get_weights()[0])

        if self.use_bias:
            self.bias = self.add_weight((self.units,),
                                     initializer=self.bias_initializer,
                                     name='bias',
                                     regularizer=self.bias_regularizer,
                                     constraint=self.bias_constraint)
        else:
            self.bias = None

        self.built = True

    def get_config(self):
        config = {'activation': activations.serialize(self.activation),
                  'other_layer': self.other_layer
                 }
        base_config = super(DenseTranspose, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))
```

**Comment by rohit-gupta** [comment]
@lvapeab I remember reading somewhere that keras models are pickle-able now. Maybe that can be a hacky solution to this replacing the need to use load_weights()

**Comment by benam2** [comment]
any update on this? how to load a model which has dense_tied layer, like the things we do in autoencoder to share the weight between encoder and decoder?

**Comment by Tr0yana** [comment]
This still seems to be an issue. Did anyone get this to work? I don't get an exception when I try this but after loading the model the accuracy is super low
